Typescript rewrite

HerbCaudill opened the pull request:
Changes referenced in https://github.com/estevanmaito/windmill-react-ui/pull/14#issuecomment-701383518; all tests passing.

estevanmaito commented on src/Button.tsx:
```diff
@@ -1,29 +1,42 @@
+  tag: 'a'
+}
+
+export type ButtonProps = ButtonAsButtonProps | ButtonAsAnchorProps
```
What about `react-router` Links?

Current usage: https://github.com/estevanmaito/windmill-dashboard-react/blob/master/src/pages/Buttons.js#L33

`react-router` supports a `component` prop, but it expects an anchor tag, and in our case we render a `button` as default. It could be used like `<Link to="/" component={Button} />` if there was a way.

gillisandrew replied:
Maybe add additional exports with pre-bound tag prop?

```ts
import React from 'react'
import { ButtonAnchor as Button } from '@windmill/react-ui'

const CustomLink: React.FC = () => <Link to="/" component={Button} />

```

Would make things a little more convenient. Otherwise the developer can wrap it themselves.

HerbCaudill replied:
resolved in 1df0f17

estevanmaito reviewed (changes requested):
I just pointed some minor issues but overall this is great!

Thank you guys for the effort!

---

### Notes to myself

- Test `Button` and `react-router` locally
- DRY `hasValidation` (repeated inside `HelperText`, `Input`, `Select` and `Textarea`)

HerbCaudill commented on src/Transition.tsx:
```diff
@@ -95,33 +96,26 @@ const CSSTransition: React.FC<TransitionProps> = function CSSTransition({
   const isChild = show === undefined
 
   if (isChild) {
+    return <CSSTransition appear={parent.appear} show={parent.show} {...rest} />
```
I wasn't sure what was going on here in Adam's original [gist](https://gist.github.com/adamwathan/e0a791aa0419098a7ece70028b2e641e#file-transition-js-L84). 

I don't totally understand what the `appear` prop is for in `CSSTransition`, but from what I can tell it's a class name, like `enter` and `exit`. It doesn't make sense to `OR` it with a boolean. 

http://reactcommunity.org/react-transition-group/css-transition

I assumed that Adam just goofed there, thinking that `appear` is a boolean - it would be an easy mistake to make. 

As a side note - I wonder if it wouldn't make more sense to just use the Transition component provided by Tailwind's new [headlessui](https://github.com/tailwindlabs/headlessui/tree/develop/packages/%40headlessui-react) project.

HerbCaudill commented on src/Windmill.tsx:
```diff
@@ -4,8 +4,7 @@ import defaultTheme from './themes/default'
 import useDarkMode from './utils/useDarkMode'
 
-interface Props {
-  children: React.ReactNode
```
👍 c95fc0a4107b875447be0de4e610b39fcfe515d9

HerbCaudill reviewed (commented):
This should be good to go - LMK if there's anything else!

HerbCaudill commented:
I've tweaked a few more things to get this building properly:

- I've added a `build:ts` step that generates the `d.ts` files, since webpack doesn't do that for us
- Tweaked the tsconfig file so that `tsc` knows which files to compile and where to put the output\
- Added a `resolve` element to the webpack config so that it knows how to resolve typescript files
- Put all the ambient type declarations in one place

estevanmaito reviewed (commented):
It looks great!

Thank you guys for the huge HUGE help!

I've decided to leave the `Input`'s `autoFocus` outside of this for now so we can better investigate it's need.

estevanmaito commented:
@HerbCaudill I tried running `npm run test:coverage` and it returns an empty list of files, as if there wasn't anything to report. Is there something I should consider now that we're using TS?

gillisandrew commented:
@estevanmaito Looks like the file matcher is not formatted correctly.
[This value](https://github.com/estevanmaito/windmill-react-ui/blob/60c747b757acaf1603e8f459002c25ee75d4c098/package.json#L139) should be something like `src/**/*.{ts,tsx}` instead of `src/**/*.tsx?$`

estevanmaito commented:
> 
> 
> @estevanmaito Looks like the file matcher is not formatted correctly.
> [This value](https://github.com/estevanmaito/windmill-react-ui/blob/60c747b757acaf1603e8f459002c25ee75d4c098/package.json#L139) should be something like `src/**/*.{ts,tsx}` instead of `src/**/*.tsx?$`

Did it work for you? I'm still getting the same results.

```
----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------|---------|----------|---------|---------|-------------------
All files |       0 |        0 |       0 |       0 |                   
----------|---------|----------|---------|---------|-------------------

Test Suites: 29 passed, 29 total  
Tests:       186 passed, 186 total
Snapshots:   0 total
Time:        8.559 s
Ran all test suites.
```

gillisandrew commented:
Yep

https://github.com/gillisandrew/windmill-react-ui/commit/2e299f36575d53a5564b118217139aa12d759bdb


```
---------------------|---------|----------|---------|---------|-------------------
File                 | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
---------------------|---------|----------|---------|---------|-------------------
All files            |   96.85 |    93.16 |   68.14 |   96.47 |
 src                 |   97.82 |    96.23 |   65.71 |   97.42 |
  Avatar.tsx         |     100 |      100 |     100 |     100 |
  Backdrop.tsx       |     100 |      100 |     100 |     100 |
  Badge.tsx          |     100 |      100 |     100 |     100 |
  Button.tsx         |     100 |      100 |     100 |     100 |
  Card.tsx           |     100 |      100 |     100 |     100 |
  CardBody.tsx       |     100 |      100 |     100 |     100 |
  Dropdown.tsx       |     100 |      100 |     100 |     100 |
  DropdownItem.tsx   |     100 |      100 |     100 |     100 |
  HelperText.tsx     |     100 |      100 |     100 |     100 |
  Input.tsx          |     100 |      100 |     100 |     100 |
  Label.tsx          |     100 |      100 |     100 |     100 |
  Modal.tsx          |     100 |      100 |     100 |     100 |
  ModalBody.tsx      |     100 |      100 |     100 |     100 |
  ModalFooter.tsx    |     100 |      100 |     100 |     100 |
  ModalHeader.tsx    |     100 |      100 |     100 |     100 |
  Pagination.tsx     |     100 |      100 |     100 |     100 |
  Select.tsx         |     100 |      100 |     100 |     100 |
  Table.tsx          |     100 |      100 |     100 |     100 |
  TableBody.tsx      |     100 |      100 |     100 |     100 |
  TableCell.tsx      |     100 |      100 |     100 |     100 |
  TableContainer.tsx |     100 |      100 |     100 |     100 |
  TableFooter.tsx    |     100 |      100 |     100 |     100 |
  TableHeader.tsx    |     100 |      100 |     100 |     100 |
  TableRow.tsx       |     100 |      100 |     100 |     100 |
  Textarea.tsx       |     100 |      100 |     100 |     100 |
  Transition.tsx     |   81.03 |    77.78 |   52.63 |   74.42 | 57-61,70-90
  Windmill.tsx       |      96 |       75 |     100 |   95.24 | 26
  index.ts           |     100 |      100 |    3.57 |     100 |
 src/context         |     100 |      100 |     100 |     100 |
  ThemeContext.tsx   |     100 |      100 |     100 |     100 |
 src/themes          |     100 |      100 |     100 |     100 |
  default.ts         |     100 |      100 |     100 |     100 |
 src/utils           |   84.44 |    77.42 |     100 |   84.62 |
  mergeDeep.ts       |   93.33 |    92.31 |     100 |     100 | 19
  useDarkMode.ts     |    87.5 |    83.33 |     100 |   85.71 | 16-18
  warning.ts         |      50 |    33.33 |     100 |      50 | 10-13
---------------------|---------|----------|---------|---------|-------------------
```

estevanmaito commented:
@gillisandrew Great, I removed `node_modules` and the `package-lock`, re-installed everything and it worked 🤦‍♂️
